airbyte-integrations/connectors/source-slack/source_slack/source.py

```diff
@@ -0,0 +1,286 @@
+
+
+class Threads(IncrementalMessageStream):
+    def __init__(self, lookback_window: Mapping[str, int], **kwargs):
+        self.messages_lookback_window = lookback_window
+        super().__init__(**kwargs)
+
+    def path(self, **kwargs) -> str:
+        return "conversations.replies"
+
+    def batches(self, stream_state: Mapping[str, Any] = None, **kwargs) -> Iterable[Optional[Mapping[str, any]]]:
+        """
+        The logic for incrementally syncing threads is not very obvious, so buckle up.
+
+        To get all messages in a thread, one must specify the channel and timestamp of the parent (first) message of that thread, basically its ID.
+
+        One complication is that threads can be updated at any time in the future. Therefore, if we wanted to comprehensively sync data i.e: get every
+        single response in a thread, we'd have to read every message in the slack instance every time we ran a sync, because otherwise there is no
+        way to guarantee that a thread deep in the past didn't receive a new message.
+
+        A pragmatic workaround is to say we want threads to be at least N days fresh i.e: look back N days into the past, get every message since,
+        and read all of the thread responses. This is essentially the approach we're taking here via batching: create batches from N days into the
+        past and read all messages in threads since then. We could optionally filter out records we have already read, but that's omitted to keep
+        the logic simple to reason about.
+
+        Good luck.
+        """
+
+        stream_state = stream_state or {}
+        channels_stream = Channels(authenticator=self.authenticator)
+        if "start_ts" in stream_state:
+            # Since new messages can be posted to threads continuously after the parent message has been posted, we get messages from the latest date
+            # found in the state minus 7 days to pick up any new messages in threads.
+            # If there is state always use lookback
+            messages_start_date = pendulum.from_timestamp(stream_state.get("start_ts")).subtract(**self.messages_lookback_window)
+        else:
+            # If there is no state i.e: this is the first sync then there is no use for lookback, just get messages from the default start date
+            messages_start_date = self._default_start_date
+        messages_stream = ChannelMessages(authenticator=self.authenticator, default_start_date=messages_start_date)
+
+        for message_chunk in messages_stream.batches(stream_state={"start_ts": messages_start_date.timestamp()}):
+            for channel in channels_stream.read_records(sync_mode=SyncMode.full_refresh):
+                message_chunk['channel'] = channel['id']
+                for message in messages_stream.read_records(sync_mode=SyncMode.full_refresh, batch=message_chunk):
+                    yield {"channel": channel['id'], "ts": message['ts']}
+
+
+class JoinChannelsStream(HttpStream):
+    """
```

Comment:
Is the purpose of this class to run with the normal sync and make sure we join all newly created channels and thus receive all new messages?

Interesting use of streams.

Reply:
This feels very strange to do within the streams framework.